src/main/java/io/specto/hoverfly/junit/core/Hoverfly.java

```diff
@@ -163,81 +164,90 @@ public void start() {
     }
 
     private void startHoverflyProcess() {
-        checkPortInUse(hoverflyConfig.getProxyPort());
-        checkPortInUse(hoverflyConfig.getAdminPort());
+        synchronized (Hoverfly.class) {
 
-        final SystemConfig systemConfig = new SystemConfigFactory(hoverflyConfig).createSystemConfig();
+            if (hoverflyConfig.getProxyPort() == 0) {
+                hoverflyConfig.setProxyPort(findUnusedPort());
+            }
+            if (hoverflyConfig.getAdminPort() == 0) {
+                hoverflyConfig.setAdminPort(findUnusedPort());
+            }
+            checkPortInUse(hoverflyConfig.getProxyPort());
+            checkPortInUse(hoverflyConfig.getAdminPort());
```

Comment:
We should move any port checking and setting right before the start of hoverfly, to avoid synchronizing the whole method, only synchronizing the following is what you wanted (in pseudocode):
```
allOtherSetup()

synchronized() {
    findAndSetPort()
    checkPortInUse()
    setPortToCommandArgs()
    startHoverfly()
}
```